[Quant][fx] Refactor lowering code (part 2)

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74619

Summary: This commit is part 2 of the effort to refactor the
lowering code in _lower_to_native_backend.py. The main change
included in this commit is generalizing the pattern matching
code across different lowering functions. There should be no
change in behavior with this PR.

A future commit will further merge the static and dynamic
lowering code paths.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

Differential Revision: [D35082210](https://our.internmc.facebook.com/intern/diff/D35082210)

andrewor14 committed:
[Quant][fx] Refactor lowering code (part 2)

Summary: This commit is part 2 of the effort to refactor the
lowering code in _lower_to_native_backend.py. The main change
included in this commit is generalizing the pattern matching
code across different lowering functions. There should be no
change in behavior with this PR.

A future commit will further merge the static and dynamic
lowering code paths.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35082210).

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -360,6 +360,89 @@ def load_arg(a):
+    matched_dequantize = False
+    for i in dequantize_node_arg_indices:
+        if i >= len(ref_node.args):
+            raise ValueError("Dequantize index %s exceeded reference node's arg length %s" % (i, len(ref_node.args)))
```
this should be an internal error I think, so it should be assert (I think we are using ValueError for user errors)

andrewor14 replied:
Ok

andrewor14 committed:
Update on "[Quant][fx] Refactor lowering code (part 2)"


Summary: This commit is part 2 of the effort to refactor the
lowering code in _lower_to_native_backend.py. The main change
included in this commit is generalizing the pattern matching
code across different lowering functions. There should be no
change in behavior with this PR.

A future commit will further merge the static and dynamic
lowering code paths.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

Differential Revision: [D35082210](https://our.internmc.facebook.com/intern/diff/D35082210)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35082210).
